engine/src/main/java/io/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java

```diff
@@ -390,36 +390,59 @@ public boolean terminateChildInstances(final BpmnElementContext context) {
 
   public void onElementCompleted(
       final ExecutableFlowElement element, final BpmnElementContext childContext) {
-    final ExecutableFlowElement containerScope;
-    final BpmnElementContext containerContext;
-    final var flowScope = element.getFlowScope();
-    if (flowScope != null) {
-      containerScope = flowScope;
-      containerContext = stateBehavior.getFlowScopeContext(childContext);
-    } else {
-      // no flowscope, assume this is called from a parent process
-      containerContext = stateBehavior.getParentElementInstanceContext(childContext);
-      containerScope = getParentProcessScope(containerContext, childContext);
-    }
-    final var containerProcessor = processorLookUp.apply(containerScope.getElementType());
-    containerProcessor.onChildCompleted(containerScope, containerContext, childContext);
+
+    invokeElementContainer(
+        element,
+        childContext,
+        (containerProcessor, containerScope, containerContext) ->
+            containerProcessor.onChildCompleted(containerScope, containerContext, childContext));
   }
 
   public void onElementTerminated(
       final ExecutableFlowElement element, final BpmnElementContext childContext) {
+
+    invokeElementContainer(
+        element,
+        childContext,
+        (containerProcessor, containerScope, containerContext) ->
+            containerProcessor.onChildTerminated(containerScope, containerContext, childContext));
+  }
+
+  public void onElementActivating(
+      final ExecutableFlowElement element, final BpmnElementContext childContext) {
+
+    invokeElementContainer(
+        element,
+        childContext,
+        (containerProcessor, containerScope, containerContext) ->
+            containerProcessor.onChildActivating(containerScope, containerContext, childContext));
+  }
+
+  private void invokeElementContainer(
```

Comment:
Optional change: 
The name does not convey that there might be no element container, in which case the method is a NOP.

```suggestion
  private void invokeElementContainerIfPresent(
```